drivers/bluetooth/hci/open_amp_nrf.c

```diff
@@ -0,0 +1,465 @@
+	k_sem_take(&sync_sem, K_FOREVER);
+
+	return 0;
+}
+
+static inline void bt_buf_set_prio(struct net_buf *buf, bool prio)
+{
+	*((u8_t *)net_buf_user_data(buf) + 1) = prio;
+}
+
+static inline bool bt_buf_get_prio(struct net_buf *buf)
+{
+	return (bool)(*((u8_t *)net_buf_user_data(buf) + 1));
+}
+
+static void rx_thread(void *p1, void *p2, void *p3)
+{
+	struct net_buf *buf;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	BT_DBG("Rx thread started");
+
+	while (1) {
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+		do {
+			BT_DBG("Calling bt_recv(%p)", buf);
+			if (bt_buf_get_prio(buf)) {
+				bt_recv_prio(buf);
+			} else {
+				bt_recv(buf);
+			}
+
+			/* Give other threads a chance to run if the ISR
+			 * is receiving data so fast that rx.fifo never
+			 * or very rarely goes empty.
+			 */
+			k_yield();
+
+			buf = net_buf_get(&rx_queue, K_NO_WAIT);
+		} while (buf);
+	}
+}
+
+static struct net_buf *bt_open_amp_evt_recv(u8_t **data, size_t *remaining)
+{
+	struct bt_hci_evt_hdr hdr;
+	struct net_buf *buf;
+
+	if (*remaining < sizeof(hdr))
+	{
+		BT_ERR("Not enought data for event header");
+		return NULL;
+	}
+
+	memcpy((void *)&hdr, *data, sizeof(hdr));
+	*data += sizeof(hdr);
+	*remaining -= sizeof(hdr);
+
+	if (*remaining != hdr.len) {
+		BT_ERR("Event payload length is not correct");
+	}
+	*remaining = hdr.len;
+	BT_DBG("len %u", hdr.len);
+
+	buf = bt_buf_get_evt(hdr.evt, false, K_NO_WAIT);
+	if (!buf) {
+		BT_ERR("No available event buffers!");
+		return buf;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	bt_buf_set_prio(buf, bt_hci_evt_is_prio(hdr.evt));
+
+	return buf;
+}
+
+static struct net_buf *bt_open_amp_acl_recv(u8_t **data, size_t *remaining)
+{
+	struct bt_hci_acl_hdr hdr;
+	struct net_buf *buf;
+
+	if (*remaining < sizeof(hdr)) {
+		BT_ERR("Not enought data for ACL header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, *data, sizeof(hdr));
+		*data += sizeof(hdr);
+		*remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+		bt_buf_set_prio(buf, false);
+	} else {
+		BT_ERR("No available ACL buffers!");
+	}
+
+	if (*remaining != hdr.len) {
+		BT_ERR("ACL payload length is not correct");
+	}
+	*remaining = sys_le16_to_cpu(hdr.len);
+
+	BT_DBG("len %u", *remaining);
+
+	return buf;
+}
+
+static void bt_open_amp_rx(u8_t *data, size_t len)
+{
+	u8_t pkt_indicator;
+	struct net_buf *buf = NULL;
+	size_t remaining = len;
+
+	BT_HEXDUMP_DBG(data, len, "Open AMP data:");
+
+	pkt_indicator = *data++;
+	remaining -= sizeof(pkt_indicator);
+
+	switch (pkt_indicator) {
+	case OPEN_AMP_EVT:
+		buf = bt_open_amp_evt_recv(&data, &remaining);
+		break;
+
+	case OPEN_AMP_ACL:
+		buf = bt_open_amp_acl_recv(&data, &remaining);
+		break;
+
+	default:
+		BT_ERR("Unknown HCI type %u", pkt_indicator);
+		return;
+	}
+
+	if (buf) {
+		BT_DBG("Remaining: %d", remaining);
+		net_buf_add_mem(buf, data, remaining);
+		net_buf_put(&rx_queue, buf);
```

Comment:
Coming back to my comment regarding the driver-internal RX thread (which wasn't answered yet, btw), my point was that why can't you just call `bt_recv()`/`bt_recv_prio()` here and forget about the rx_queue and a separate thread? `bt_recv()` will take care of deferring to the host-side RX thread since that gets enabled due to the way this driver is defined in Kconfig.

Reply:
I agree. (commented longer answer in the original comment that wasn't answered.).

Reply:
Sorry for the late response, but I was offline

@jhedberg, @joerchan 
Yes, it seems that the RX thread is redundant in this case. I was too heavily inspired by h4 implementation of HCI driver, which contained a lot more processing logic (HCI packet assembly, etc.).

I also missed the fact that the context switching (to another (Host) thread) happens right after bt_recv() call like @joerchan pointed out:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L4866

I`m going to send a patch fixing it in a second.

Reply:
done, please take a look